modape/modis/download.py

```diff
@@ -234,55 +270,72 @@ def download(self,
-                # warm up pool
-                _ = session.get(self.results[0]["link"], stream=True, allow_redirects=True)
+            with SessionWithHeaderRedirection(username, password) as session:
 
-                with ThreadPoolExecutor(nthreads) as executor:
+                retries = Retry(total=5, backoff_factor=1, status_forcelist=[502, 503, 504])
+                session.mount(
+                    "https://",
+                    HTTPAdapter(pool_connections=nthreads, pool_maxsize=nthreads*2, max_retries=retries)
+                )
 
-                    futures = [executor.submit(self._fetch_hdf, session, x["link"], targetdir, overwrite)
-                               for x in self.results]
+                if multithread:
+                    log.debug("Multithreaded download using %s threads. Warming up connection pool.", nthreads)
+                    # warm up pool
+                    _ = session.get(list(to_download.values())[0]["link"], stream=True, allow_redirects=True)
 
-                downloaded_files = [x.result() for x in futures]
+                    with ThreadPoolExecutor(nthreads) as executor:
 
-            else:
-                log.debug("Serial download")
-                downloaded_files = []
+                        futures = [executor.submit(self._fetch, session, values["link"], targetdir, overwrite, robust)
+                                   for key, values in to_download.items()]
 
-                for result in self.results:
+                    downloaded_temp = [x.result() for x in futures]
 
-                    downloaded_files.append(
-                        self._fetch_hdf(session, result["link"], targetdir, overwrite)
-                    )
+                else:
+                    log.debug("Serial download")
+                    downloaded_temp = []
 
-        errors = []
+                    for _, values in to_download.items():
 
-        # check if downloads are OK
-        for file, err in downloaded_files:
+                        downloaded_temp.append(
+                            self._fetch(session, values["link"], targetdir, overwrite, robust)
+                        )
 
-            if err is not None:
-                errors.append((file, err)) # append to error list
-            else:
-                # if no error, make sure file is on disk
-                assert file.exists(), "Downloaded file is missing! No download error was reported"
+            # check if downloads are OK
+            for fid, err in downloaded_temp:
+                if err is None:
+                    del to_download[fid]
+                    downloaded.append(fid)
+
+            if to_download:
+                n_todo = len(to_download)
+                if retry_count < max_retries or max_retries == -1:
+                    retry_count += 1
+                    log.debug("Retrying downloads! Files left: %s", n_todo)
```

Comment:
`n_todo` can be substituted by inlining the `len(to_download)` statement (line 329)

Reply:
right, this is a remainder from a previous version.